pkg/sql/colexec/external/external.go

```diff
@@ -0,0 +1,820 @@
+					nulls.Add(vec.Nsp, uint64(rowIdx))
+				} else {
+					d, err := types.Decimal64_FromString(field)
+					if err != nil {
+						// we tolerate loss of digits.
+						if !moerr.IsMoErrCode(err, moerr.DATA_TRUNCATED) {
+							logutil.Errorf("parse field[%v] err:%v", field, err)
+							return nil, fmt.Errorf("the input value '%v' is not Decimal64 type for colnum %d", field, colIdx)
+						}
+					}
+					cols[rowIdx] = d
+				}
+			case types.T_decimal128:
+				cols := vec.Col.([]types.Decimal128)
+				if isNullOrEmpty {
+					nulls.Add(vec.Nsp, uint64(rowIdx))
+				} else {
+					d, err := types.Decimal128_FromString(field)
+					if err != nil {
+						// we tolerate loss of digits.
+						if !moerr.IsMoErrCode(err, moerr.DATA_TRUNCATED) {
+							logutil.Errorf("parse field[%v] err:%v", field, err)
+							return nil, fmt.Errorf("the input value '%v' is not Decimal128 type for colnum %d", field, colIdx)
+						}
+					}
+					cols[rowIdx] = d
+				}
+			case types.T_timestamp:
+				cols := vec.Col.([]types.Timestamp)
+				if isNullOrEmpty {
+					nulls.Add(vec.Nsp, uint64(rowIdx))
+				} else {
+					d, err := types.ParseTimestamp(field, vec.Typ.Precision)
+					if err != nil {
+						logutil.Errorf("parse field[%v] err:%v", field, err)
+						return nil, fmt.Errorf("the input value '%v' is not Timestamp type for colnum %d", field, colIdx)
+					}
+					cols[rowIdx] = d
+				}
+			default:
+				return nil, fmt.Errorf("the value type %d is not support now", param.Cols[rowIdx].Typ.Id)
+			}
+		}
+	}
+	n := vector.Length(bat.Vecs[0])
+	sels := proc.Mp.GetSels()
+	if n > cap(sels) {
+		proc.Mp.PutSels(sels)
+		sels = make([]int64, n)
+	}
+	bat.Zs = sels[:n]
+	for k := 0; k < n; k++ {
+		bat.Zs[k] = 1
+	}
+
+	return bat, nil
+}
+
+// read batch data from external file
+func GetSimdcsvReader(param *ExternalParam) (*ParseLineHandler, error) {
+	var err error
+	param.reader, err = getLoadDataReader(param.load)
+	if err != nil {
+		logutil.Errorf("open file failed. err:%v", err)
+		return nil, err
+	}
+	param.reader, err = getUnCompressReader(param.load, param.reader)
+	if err != nil {
+		return nil, err
+	}
+
+	channelSize := 100
+	plh := &ParseLineHandler{}
+	plh.batchSize = param.batchSize
+	plh.simdCsvGetParsedLinesChan = atomic.Value{}
+	plh.simdCsvGetParsedLinesChan.Store(make(chan simdcsv.LineOut, channelSize))
+	if param.load.Tail.Fields == nil {
+		param.load.Tail.Fields = &tree.Fields{Terminated: ","}
+	}
+	plh.simdCsvReader = simdcsv.NewReaderWithOptions(param.reader,
+		rune(param.load.Tail.Fields.Terminated[0]),
+		'#',
+		false,
+		false)
+
+	return plh, nil
+}
+
+// read batch data from external file
+func ScanFileData(param *ExternalParam, proc *process.Process) (*batch.Batch, error) {
+	var bat *batch.Batch
+	var err error
+	if param.plh == nil {
+		param.plh, err = GetSimdcsvReader(param)
+		if err != nil {
+			return nil, err
+		}
+	}
+	plh := param.plh
+	plh.simdCsvLineArray, err = plh.simdCsvReader.Read(param.batchSize)
```

Comment:
Where is the read implementation? I can't find the corresponding code implementation

Reply:
it is now in my branch code, I will push it later